### PR TITLE
Preliminary PR on optim-wip

### PR DESCRIPTION
**Pull request by greentfrapp**
Submitting this somewhat raw PR after an initial re-org of the code, primarily to facilitate collaboration with @ProGamerGov.

Changes made:
- Re-organized code as per our discussion
  - Also added optim._utils.typing for new types, we can shift this elsewhere as well
- Added initial implementation of ImageTensor wrapper around torch.Tensor
- Added quick tutorial in tutorials/TorchVision_OptimViz.ipynb
- Removed unnecessary imports and linted

### CODE REVIEW

**Comment by greentfrapp** [comment]
Note to self: isort == 4.3.21 sorts differently form isort == 5.6.4.
Specifically in 4.3.21, `from` imports belong in their own section (preceded by a newline), while in 5.6.4 this is not true.

**Comment by NarineK** [comment]
Thank you very much @greentfrapp ! I'll take a look.

File: captum/optim/_param/image/images.py [hunk 1]
```diff
@@ -254,10 +307,9 @@ def forward(self):
         image = self.parameterization()
         image = self.decorrelate(image)
         image = image.rename(None)  # TODO: the world is not yet ready
-        return torch.sigmoid_(image)
+        return CudaImageTensor(torch.sigmoid_(image))
```
Comment (NarineK) [inline captum/optim/_param/image/images.py:310]: Does this have to be `CudaImageTensor` ? Can't it be regular `ImageTensor` that can be on any device ?

File: captum/optim/_core/objectives.py [hunk 2]
```diff
@@ -0,0 +1,261 @@
...
+
+        if self.transform:
+            image = self.transform(image)
+
+        with suppress(AbortForwardException):
+            _unreachable = self.model(image)  # noqa: F841
+
+        # consume_outputs return the captured values and resets the hook's state
+        module_outputs = self.hooks.consume_outputs()
+        loss_value = self.loss_function(module_outputs)
```
Comment (NarineK) [inline captum/optim/_core/objectives.py:80]: Can we pass `targets` to `loss_function` here so that the user doesn't have to pass `targets` to `channel_activation` or `neuron_activation` when initializing it ? Or perhaps we can do that in the constructor ?

File: captum/optim/_core/objectives.py [hunk 3]
```diff
@@ -0,0 +1,261 @@
...
+        # assert len(activations.shape) == 4
+        return activations[:, channel_index, ...]
+
+    return loss_function
+
+
+def neuron_activation(
+    target: nn.Module, channel_index: int, x: int = None, y: int = None
+) -> LossFunction:
+    # ensure channel_index will be valid
```
Comment (NarineK) [inline captum/optim/_core/objectives.py:176]: This and channel_activation seem also image specific
To make this more generic we can call `channel_index`, `neuron_selector` which can also be a callable function that performs 
```
activations[:, channel_index, ...]
```
for example. Can `neuron_selector` Callable argument help us to generalize `channel_activation` and  `neuron_activation ` under one umbrella ?

**Review by NarineK — APPROVED** [review]
Looks great! Thank you for working on this @greentfrapp ! I left couple comments. There are some places in _core that are still image specific that we can generalize also in later PR. I also made couple suggestions for the API.

File: captum/optim/__init__.py [hunk 4]
```diff
@@ -1,27 +1,9 @@
...
-
-ModuleOutputMapping = Dict[nn.Module, Optional[torch.Tensor]]
-
-StopCriteria = Callable[[int, Objective, torch.optim.Optimizer], bool]
+import captum.optim._core.objectives as objectives  # noqa: F401
+import captum.optim._param.image.images as images  # noqa: F401
+import captum.optim._param.image.transform as transform  # noqa: F401
+import captum.optim._utils.typing as typing  # noqa: F401
 
+from ._core.objectives import InputOptimization  # noqa: F401
```
[the remark this answers is not in the document]
Comment (greentfrapp) [reply captum/optim/__init__.py]: Yup fixed!

File: captum/optim/_core/objectives.py [hunk 5]
```diff
@@ -0,0 +1,261 @@
...
+    Parameterized,
+    SingleTargetLossFunction,
+    StopCriteria,
+)
+from .output_hook import AbortForwardException, ModuleOutputsHook
+
+
+class InputOptimization(Objective, Parameterized):
+    model: nn.Module
+    input_param: ImageParameterization
```
[the remark this answers is not in the document]
Comment (greentfrapp) [reply captum/optim/_core/objectives.py]: Added InputParameterization and ImageParameterization extends from InputParametrization.

File: captum/optim/_core/objectives.py [hunk 6]
```diff
@@ -0,0 +1,261 @@
...
+    Objective,
+    Parameterized,
+    SingleTargetLossFunction,
+    StopCriteria,
+)
+from .output_hook import AbortForwardException, ModuleOutputsHook
+
+
+class InputOptimization(Objective, Parameterized):
+    model: nn.Module
```
[the remark this answers is not in the document]
Comment (greentfrapp) [reply captum/optim/_core/objectives.py]: Yup, removed these and using the typehints in the constructor instead.

File: captum/optim/_core/objectives.py [hunk 7]
```diff
@@ -0,0 +1,261 @@
...
+        https://github.com/tensorflow/lucid
+        https://distill.pub/2017/feature-visualization/
+    """
+
+    def __init__(
+        self,
+        model: nn.Module,
+        input_param: Optional[nn.Module],
+        transform: Optional[nn.Module],
+        targets: Iterable[nn.Module],
```
[the remark this answers is not in the document]
Comment (greentfrapp) [reply captum/optim/_core/objectives.py]: Sounds good, changed!

File: captum/optim/_param/image/transform.py [hunk 8]
```diff
@@ -214,6 +219,16 @@ def forward(self, input):
         return self.conv(input, weight=self.weight, groups=self.groups)
 
 
+class Normalize(nn.Module):
```
[the remark this answers is not in the document]
Comment (greentfrapp) [reply captum/optim/_param/image/transform.py:222]: Yup I thought about this. The way Ludwig set this up in his initial commit, the transforms need to extend from torch.nn.Module and the torchvision Normalize does not do that. I'll leave this first but we can see if there's a better way to integrate this with torchvision transforms.

File: captum/optim/_param/image/transform.py [hunk 9]
```diff
@@ -214,6 +219,16 @@ def forward(self, input):
...
+        super().__init__()
+        self.mean = torch.as_tensor(mean).view(3, 1, 1).to(device)
+        self.std = torch.as_tensor(std).view(3, 1, 1).to(device)
+
+    def forward(self, x):
+        return (x - self.mean) / self.std
+
+
 def test_transform():
     from clarity.pytorch.fixtures import image
```
[the remark this answers is not in the document]
Comment (greentfrapp) [reply captum/optim/_param/image/transform.py:233]: I'm leaving transform.py largely untouched, since Ben is probably going to change quite a bit of it.

File: captum/optim/_param/image/images.py (hunk 1 above)
Comment (greentfrapp) [reply captum/optim/_param/image/images.py:315]: I think we can have a regular ImageTensor, but I had some issues when switching between cuda and cpu. Will need some time to figure it out, will merge this for the time being.

File: captum/optim/_core/objectives.py (hunk 2 above)
Comment (greentfrapp) [reply captum/optim/_core/objectives.py:77]: Yup i agree with this, a little redundant to be passing same targets twice. I'll look into this and figure out a neater implementation in a future PR.

File: captum/optim/_param/image/transform.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (ProGamerGov) [reply captum/optim/_param/image/transform.py:222]: @NarineK The tensor being optimized is kept in a untransformed form and then the preprocessing transforms are performed right before it's sent through the network. The sigmoid function used after recorrelation only outputs 0-1 and thus normalizing the tensor before that will erase the negative values.
